### ISSUE TITLE
Unable to provision node driver cluster using CentOS

### ISSUE THREAD

**Issue by rishabhmsra** [rancher/rancher#36820]
**Rancher Server Setup**
- Rancher version: v2.6-head(47c2238)
- Installation option (Docker install/Helm Chart): docker
   
**Information about the Cluster**
- Kubernetes version: v1.23.4-rancher1-1
- Cluster Type (Local/Downstream): 
   - If downstream, what type of cluster? (Custom/Imported or specify provider for Hosted/Infrastructure Provider): ec2 node driver

**To Reproduce**
- Provision a ec2 node driver cluster(single node) -> using a node template having CentOS ami. In this case CentOS 8.3 was used. The cluster will stuck into provisioning state with message : `Waiting for etcd, controlplane and worker nodes to be registered`
- **Rancher Server Logs :** 
```
2022/03/09 12:45:27 [INFO] [node-controller-rancher-machine] Detecting operating system of created instance...
2022/03/09 12:45:27 [INFO] [node-controller-rancher-machine] Waiting for SSH to be available...
2022/03/09 12:46:26 [INFO] [node-controller-rancher-machine] Detecting the provisioner...
2022/03/09 12:46:31 [INFO] [node-controller-rancher-machine] Provisioning with centos...
2022/03/09 12:46:53 [INFO] [node-controller-rancher-machine] The default lines below are for a sh/bash shell, you can specify the shell you're using, with the --shell flag.
2022/03/09 12:46:53 [INFO] [node-controller-rancher-machine] 
2022/03/09 12:46:53 [INFO] Generating and uploading node config rishabh-centos-docker1
2022/03/09 12:46:53 [ERROR] error syncing 'c-42m8s/m-x9n9p': handler node-controller: Error creating machine: RHEL ssh command error: command: sudo -E yum install -y curl erCentOS Linux 8 - AppStream                      118  B/s |  38  B     00:00    , requeuing     --:-- ETA
2022/03/09 12:48:54 [INFO] Removing node rishabh-centos-docker1
```

- **OS Details:**
```
NAME="CentOS Linux"
VERSION="8"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="8"
PLATFORM_ID="platform:el8"
PRETTY_NAME="CentOS Linux 8"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:8"
HOME_URL="https://centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"
CENTOS_MANTISBT_PROJECT="CentOS-8"
CENTOS_MANTISBT_PROJECT_VERSION="8"
```
```
cat /etc/centos-release
CentOS Linux release 8.3.2011
```
**Expected Result**
- Downstream cluster should come into Active state.

**Additional Context**
https://github.com/rancher/rancher/issues/34615#issuecomment-1058085932

**Comment by paynejacob** [comment]
This is likely an issue in rancher-machine.  The error message indicates that the curl install is failing.  The centos package registries have changed recently and likely need to be updated.  

There was a similar problem with the docker install scripts.
https://github.com/rancher/install-docker/pull/91

**Comment by snasovich** [comment]
@rishabhmsra , is this affecting both RKE1 and RKE2 provisioning?
Also, could you check if it's happening in 2.6.3?

**Comment by rishabhmsra** [comment]
@snasovich, I validated this on 2.6.3 and on v2.6-head(bab14a3), by provisioning single node, ec2 node driver cluster.

On v2.6-head:
- RKE2 cluster(v1.21.10+rke2r1) provisioned successfully.
- RKE1 cluster(v1.23.4-rancher1-1), failed to provision with same error.

On 2.6.3:
- RKE2 cluster(v1.21.9+rke2r1) provisioned successfully.
- RKE1 cluster(v1.22.6-rancher1-1) failed to provision with same error.

**Comment by HarrisonWAffel** [comment]
I'm adding this test template for the future, this issue should not be tested until the updated `install-docker` script has been released and rancher-machine has been bumped. 

### Root cause
  The publicly available `install-docker` script did not include the logic to update the repository mirrors, causing `sudo yum install` to fail. Additionally, rancher-machine was installing prerequisite packages (such as cURL) before installing docker, and thus before the repository mirrors were updated by the install-docker script. 
### What was fixed, or what changes have occurred
  The `install-docker` script has been updated to include the repository mirror fix, and rancher-machine will not install docker before prerequisite packages when provisioning centOS nodes. 
### Areas or cases that should be tested
  The creation of nodes using a centOS ami. 
### What areas could experience regressions?
  The creation of other redhat ami's (very unlikely)
  + Create a new RKE1 node template using a centOS ami, such as `ami-056b03dba13a2c9dd`
  + Create a cluster using this template and observe the node(s) go into an `active` state after being provisioned by rancher-machine

**Comment by snasovich** [comment]
@HarrisonWAffel , with the rancher-machine PR merged, are we waiting on release of updated `install-docker` script? Is there another issue PR/issue we can link here and move to Blocked as it's really not in Review?

**Comment by HarrisonWAffel** [comment]
Yes, the install-docker script that is currently in the dev bucket needs to be promoted to the production bucket. Seb has created a work flow to do this automatically upon creation of a tag. I just need to have EIO backup the bucket in case anything goes wrong and they need to restore it. I plan on having this done today so the ticket can move forward

**Comment by HarrisonWAffel** [comment]
The updated install-docker script has been promoted to the public URL ([drone build](https://drone-pr.rancher.io/rancher/install-docker/96)) and as such this ticket is ready for test

### Root cause
  The provisioning of centOS machines was broken due to two reasons
1) The `install-docker` script did not include the logic for updating repository mirrors. Because centOS is not EOL, the default mirrors no longer work and `yum install` would fail. 

2) When provisioning a centOS machine, supporting packages were installed before docker. This is normally OK as we do not expect the need to execute any prerequisite commands in order to install packages. However, for centOS we need the mirrors to be updated before _any_ installs can occur. 

### What was fixed, or what changes have occurred
  1) The `install-docker` script has since been updated to modify the repository mirrors correctly so that `yum install` works as expected.

 2) The logic for creating centOS machines has been updated to first install docker and adjust the repository mirrors before installing any other supporting packages. 
### Areas or cases that should be tested
  The creation of centOS machines. 

Test steps: 
1. Create a new RKE1 `NodeTemplate` with the following ami and SSH username 
  AMI: `ami-056b03dba13a2c9dd`, username: `centos`
2. Use this new `NodeTemplate` to create a new cluster
3. Observe the successful creation of the centOS machine, and that the cluster goes into an `active` state. 


### What areas could experience regressions?
  The provisioning of other redhat AMI's, however this is unlikely due to the implementation.

**Comment by rishabhmsra** [comment]
Validated on rancher v2.6-head(bca5ada).

Steps followed:
- Created a RKE1 node template using the ami : `ami-056b03dba13a2c9dd` and username : `centos`.
- Provisioned a single node, ec2 node driver cluster using k8s version v1.23.6-rancher1-1.
- Provisioned a multi-node(1-cp, 1-etcd, 1-w), ec2 node driver cluster using k8s version v1.23.6-rancher1-1.

Result:
- The clusters provisioned successfully and came into Active state, hence closing the issue.
